### ISSUE TITLE
borg binary fails on OSX 10.9 (prebuilt binary)

### ISSUE THREAD

**Issue by arctica** [borgbackup/borg#629]
I have installed borg backup via homebrew on OSX which worked but when executing the binary it errors out. OSX version is 10.9.2 (Mavericks).

```
$ brew install --force-bottle Caskroom/cask/borgbackup
==> Tapping caskroom/cask
...
==> Downloading https://github.com/borgbackup/borg/releases/download/0.30.0/borg-macosx64
######################################################################## 100.0%
==> Verifying checksum for Cask borgbackup
==> Symlinking Binary 'borg-macosx64' to '/usr/local/bin/borg'
🍺  borgbackup staged at '/opt/homebrew-cask/Caskroom/borgbackup/0.30.0' (5.3M)
```

```
$ borg
Error loading Python lib '/var/folders/fx/s9m97pmj1xq0cf3q_ddbm4t00000gn/T/_MEIMvpYkG/Python': dlopen(/var/folders/fx/s9m97pmj1xq0cf3q_ddbm4t00000gn/T/_MEIMvpYkG/Python, 10): Symbol not found: _faccessat
  Referenced from: /var/folders/fx/s9m97pmj1xq0cf3q_ddbm4t00000gn/T/_MEIMvpYkG/Python
  Expected in: /usr/lib/libSystem.B.dylib
 in /var/folders/fx/s9m97pmj1xq0cf3q_ddbm4t00000gn/T/_MEIMvpYkG/Python
```

Looking for this issue I've found that someone ran into the same problem on OSX trying to run the emacs mac port and there the issue resulted from building the binary on a 10.10 machine which makes the binary depend on the _faccessat symbol which isn't there on 10.9.

https://github.com/railwaycat/emacs-mac-port/issues/63

On the download page I then saw this remark for prebuilt binaries:
> Mac OS X: 10.10 (unknown whether it works for older releases)

So I guess that at least settles the "unknown" part. But it would be great if someone could build a binary that works on 10.9.

Thanks

**Comment by ThomasWaldmann** [comment]
Ah, ok, I'll update the docs.

Problem is that I only have that 10.10 machine, so I can't build on anything older.

**Comment by steveschow** [comment]
an older machine is not needed to build something that will run on older OS.  its just a configuration issue, you can configure a newer one to build stuff that will run on Mavericks or Lion, no problem.  If you have any build  instructions I will try to build it on my Mavericks system as I really want to use borg and I use Mavericks

**Comment by enkore** [comment]
Vagrant is used, here are the shell scirpts:

https://github.com/borgbackup/borg/blob/master/Vagrantfile#L52
https://github.com/borgbackup/borg/blob/master/Vagrantfile#L193

**Comment by steveschow** [comment]
I misunderstood.  I thought the 'binary" mentioned needed to be compiled.  This appears to just be question of installing all the right dependencies for the python to run?  It sounds like the "binary" that is mentioned in the docs is actually not a binary, its a vagrant bootstrap file that installs a bunch of stuff, rather then doing it manually. yes?  Why doesn't it work before Yosemite?  what is broken?

**Comment by enkore** [comment]
The binary is build with pyinstaller (also in the script). Vagrant is used for building it, not on the users' system.

**Comment by arctica** [comment]
Vagrant as far as I can tell is used to compile/build the borg binary inside a VM.

The "binary" is more or less a binary (basically a package of all the python etc into a single file) which due to being built on 10.10+ relies on having the _faccessat symbol in the libSystem.B.dylib library which isn't present on 10.9 systems.

Now if you'd run the Vargant scripts on a 10.9 system, hopefully that would result in a binary that doesn't have that dependency and can run on all versions.

**Comment by steveschow** [comment]
I see. I've never worked with vagrant before, but if I can figure that out I will do it and let you know.

**Comment by steveschow** [comment]
Did anyone ever find a way to build the borg binary for OSX 10.9?  I really really need it now.

**Comment by ThomasWaldmann** [comment]
@steveschow did you try the stuff from the Vagrantfile?

Here are some updated links for 1.0.11:

https://github.com/borgbackup/borg/blob/1.0.11/Vagrantfile#L67

https://github.com/borgbackup/borg/blob/1.0.11/Vagrantfile#L227

https://github.com/borgbackup/borg/blob/1.0.11/Vagrantfile#L303

**Comment by ThomasWaldmann** [comment]
And if you know how to modify these commands, so that the built-on-10.10 binary works on < 10.10, I'ld like to modify the Vagrantfile and have the result tested.

**Comment by steveschow** [comment]
I tried to mess with vagrant last year when I first asked about this, but I never could get it to work, I am trying again now.  

I can follow the instructions to use git src 1.1-maint and I can setup a virtualenv that will run on my mavericks machine with python3, and it seems to work...  

the binary would be more convenient though so that I don't have to worry about python3 on this machine.

I may eventually try to upgrade this machine to yosemite, just for this...but...its not trivial for reasons I won't go into now, so probably can't happen any time soon.

**Comment by ThomasWaldmann** [comment]
Vagrant is not the problem here.
But rather that nobody of us knows how to build such a old-compatible binary on 10.10.

**Comment by steveschow** [comment]
Generally its not that hard, you just set the target to an older version, but it might be that an older version SDK needs to be present, I'm not sure.  I thought that whole point of Vagrant is that it creates a virtual environment in order to build anything from anywhere...??  Isn't it just a matter of getting the right "vm"?

**Comment by steveschow** [comment]
Yea, ok, so reading the pyinstaller docs and I see the following...


> The output of PyInstaller is specific to the active operating system and the active version of Python. This means that to prepare a distribution for:
> 
> a different OS
> a different version of Python
> a 32-bit or 64-bit OS

> you run PyInstaller on that OS, under that version of Python. The Python interpreter that executes PyInstaller is part of the bundle, and it is specific to the OS and the word size.

So there it is...  can't change a target like you can when building C/C++ apps.  pyinstaller appears to be hard coded to take whatever the current environment looks like, including the OS version...and bakes that into it.  Though, at some low level, they are choosing to set the target I mentioned, as part of the pyinstaller process.

If you were to use 10.7 or 10.8 as the vagrant vm, then I think the result would work all the way through today.  I see at least one 10.8 vm on the vagrant site that looks like the right thing.  Unfortunately I don't know much about vagrant and its too much for me to figure out how to do that.  I am going to see if I can run pyinstaller though.

**Comment by steveschow** [comment]
building the binary with pyinstaller was pretty easy, just had to slightly tweak the borg.exe.spec file.  

I have a working binary.  1.1rc4

I wish you guys would consider using an older VM with vagrant, such as maybe the 10.8 one, so that future releases will be good going back a bit further.  I realize almost nobody runs 10.9 or especially 10.8 on their mac anymore, but in this case its an old mac being used as a file server...which is exactly the kind of situation where borg is useful.

**Comment by ThomasWaldmann** [comment]
can you check whether the 10.9 binary also works on latest OS X?

**Comment by steveschow** [comment]
Yes it runs on 10.12.   I will email it to you so you can try it.

**Comment by ThomasWaldmann** [comment]
I built a new binary on 10.10 (I don't want to use an older build system, if avoidable). but with setting this before building all the stuff:
```
export MACOSX_DEPLOYMENT_TARGET="10.4"
```
https://paste.thinkmo.de/JRyxwkVE#borg-macosx64-10.4

@steveschow can you test whether that works on 10.9 and 10.12? It passed the tests on 10.10.

**Comment by steveschow** [comment]
Very cool, will test shortly…

**Comment by steveschow** [comment]
Doesn’t work.  It will run..nothing stops it from running…  And if I call it with no arguments, it displays the default help usage text.  However, provide any arguments at all and seg fault.  Such as 

 ./borg-macosx64-10.4 help
Fatal Python error: Segmentation fault

Current thread 0x00007fff75293310 (most recent call first):
  File "shutil.py", line 399 in _rmtree_safe_fd
  File "shutil.py", line 480 in rmtree
  File "tempfile.py", line 812 in cleanup
  File "tempfile.py", line 808 in __exit__
  File "borg/testsuite/hashindex.py", line 303 in _serialize_hashindex
  File "borg/testsuite/hashindex.py", line 324 in test_identical_creation
  File "unittest/case.py", line 605 in run
  File "unittest/case.py", line 653 in __call__
  File "unittest/suite.py", line 122 in run
  File "unittest/suite.py", line 84 in __call__
  File "unittest/suite.py", line 122 in run
  File "borg/selftest.py", line 64 in selftest
  File "borg/archiver.py", line 3849 in prerun_checks
  File "borg/archiver.py", line 3887 in run
  File "borg/archiver.py", line 3991 in main
  File "borg/__main__.py", line 2 in <module>
Segmentation fault: 11

**Comment by steveschow** [comment]
That seg fault was on 10.9.  Works fine on 10.12, I inited a repo and ran create and list on it successfully

**Comment by steveschow** [comment]
PS - a lot changed in 10.7, I believe that is where OSX switched to full 64bit if I recall or something like.  Maybe try setting the target to 10.7?

**Comment by steveschow** [comment]
Actually never mind my last comment…  It runs fine on 10.10 and 10.12…so the problem isn’t that you set it to 10.4.  The problem is that the MACOSX_DEPLOYMENT_TARGET doesn’t appear to have any effect on what pyinstaller does.

**Comment by ThomasWaldmann** [comment]
Well, it can't be the pyinstaller bootloader (you would not see borg's help if that piece crashes) and everything else is freshly built with the target setting.

So I'll just try again with 10.7 target. Will take a bit, it will re-initialize the whole vagrant machine / rebuild all compiled stuff for that target (well, the stuff we compile, not sure if that's all going into the binary).

**Comment by steveschow** [comment]
One interesting question I have about pyinstaller, I’ve never used it before this came up.  But the question is whether the binary built by pyinstaller completely sandboxes what is inside or will the running python script also potentially attempt to make use of things that might be installed in the system related to python?  Obviously it bundles up python itself and supposedly any needed extensions, etc..  But the question is…is it completely sandboxed?

So the seg fault I’m seeing is Python itself it would seem.  whatever version of Python that your compile bundles into the binary…that won’t run on 10.9, but will on 10.10+.  

Or maybe..just a shot in the dark….your newer build of python present on the 10.10 system is not happy with the 10.4 targeted boot loader in some way, so its worth a try to see if 10.7 will work.   Might even work with 10.6, but 10.7 is safer…

**Comment by steveschow** [comment]
Just as a for example, note the mac install notes from this python release page

https://www.python.org/downloads/release/python-278/ <https://www.python.org/downloads/release/python-278/>

Notice how there is a different installer for 10.6+

There was a lot of OSX kaos during 10.4-10.5, 

So anyway, if the python3  you had on the 10.10 machine somehow needs 10.6+, then maybe the bootloader needs to also be built on that target…  its worth a try..

**Comment by ThomasWaldmann** [comment]
I've found another issue: i *thougtn* i set the env var early, but it was after installing all the brew packages and installing pyenv itself.

For the next try, with 10.7, I now really set the env var early, let's see...

**Comment by ThomasWaldmann** [comment]
Hmm, seems it pulls binary stuff from brew:
```
==> darwin64: ==> Pouring lz4-1.8.0.yosemite.bottle.tar.gz
```
And the lz4 libs *will* get bundled into the binary.

**Comment by steveschow** [comment]
That isn’t neccessarily a problem….

homebrew bundles up stuff into bottles named by OS version but it doesn’t necessarily mean the libraries are any different underneath per say..   


Also…  that particular thing is a system library….  Its not my impression that pyinstaller bundles that into the binary..it bundles in python stuff….right?  I would think the python stuff dynamically links to whatever is installed on the system at runtime.

homebrew is to mac what apt-get is to linux…

**Comment by ThomasWaldmann** [comment]
lz4, lzma, openssl/libcrypto, python 3.5.4 will go into the pyinstaller-made binary.

**Comment by steveschow** [comment]
Right, but what does into the binary is the python wrappers around those things yes?  Not the actual system libraries.

**Comment by steveschow** [comment]
I can’t find it right now, but I read online  at python.org <http://python.org/> somewhere a few minutes ago, that current python is built with a target of 10.6.  I also read some notes about 10.7+, bla bla…no mention of any OS versions newer then that.  I think 10.7 is a safe target and possibly 10.6 might work too.

**Comment by ThomasWaldmann** [comment]
No, I really mean the lz4, lzma, libcrypto libs. pyinstaller bundles that all, the only thing not bundled is the system libc as that needs to match the system.

**Comment by ThomasWaldmann** [comment]
Hmm, installing that stuff from src broke all the python builds.
Another try with binary bottles...

**Comment by steveschow** [comment]
Interesting.  well we’ll see what happens..…   

Based on the fact that the 10.4 target ran on 10.12 but not on 10.9, I think i’m ruling out the target itself as a reason for the seg fault.  But never know…its worth a try.  like I said, there were a lot of changes around 10.4-10.5 related to 64bit…  

That being said…it ran on 10.12 seemingly ok…so…I kind of doubt that is it.

The fact that homebrew installs those dependencies, including python itself…based on OS version…and its a bit of a black box to know what exactly homebrew installed.  Probably could track it down with some research I guess.. 

What I built with pyinstaller by hand on my 10.9 box seemed to run on 10.12.  That could simply be that homebrew(10.9) installed things which are forward compatible, while homebrew (10.10) installed things which are not backwards compatible.  Hard to say exactly and there is no guarantee that homebrew will always install things that are forward compatible.   

Now we know why pyinstaller is documented as only supporting binaries built on exactly the same system they will be run on.   That creates kind of a logistical nightmare for you to build mac borg binaries..you wouldn’t want to have to build for every OS version.  It may be that you can build something based on 10.9 (as I did), or 10.8 or even 10.7 VM’s that might very well work on 10.12, but according to the pyinstaller docs…this is not supported.  Through testing you might determine that it works ok, and roll with it.  But eventually some version of OSX will come out that the old build won’t work anymore and you’ll have to update the VM to something newer, or maintain 2+ VM’s for the mac platform.

Just thinking out loud here..  I think to really determine deterministically, you’d need to know more exactly what is installed by homebrew on each version of OSX and then check all compabilities and make a deterministic decision about what VM’s you need to include to support various versions of OSX.  I think its quite likely that something built on a 10.7 VM will run on 10.12, for whatever its worth but like I said, that should not be considered supported without more information about what homebrew has installed.

I would think that the exact same situation as above applies to linux distributions, by the way….jessy vs wheezy, vs whatever comes next.  

Alternatively, if you include an easy to run manual build script in the source to build the binary without vagrant, then anyone at all can follow the git install instructions, get it all working and then finally build the binary for themselves on their platform.  But ironically, most people looking for a built binary are probably trying to avoid installing python, etc to begin with.

**Comment by steveschow** [comment]
ps I am packing up for a trip to mexico leaving in a few hours, so if you post something for me to test for you, I may not get to it until after Oct 8th.

**Comment by ThomasWaldmann** [comment]
Just setting the deployment target to 10.7 early is enough to make all the Python builds fail.

Guess I'll give up on that and just stay with the normal 10.10+ build as we had it since long.

If someone wants to experiment with 10.9 machine based builds, first anwer the question whether the generated binary includes stuff (from apple or elsewhere) that does not get security updates any more.

**Comment by steveschow** [comment]
Yea I don’t think building python from scratch with the old target was the right way to go either because of the other homebrew dependencies.   I think you want python and all other python dependencies as a group compatible with each other. 

If you want to support older OSX binary, I think you’d need to use older vagrant VM.  Sooner or later in the future, quite possibly 10.13 since it involves a new file system, you will have a new hurdle where stuff will break on that one and you’ll have to change your VM to 10.13 or maintain two VM’s for 10.10-10.12 and 10.13+.  I still do think its very likely if you setup a 10.7 or 10.8 VM, the result will work 10.7-10.12, instead of the VM you’re using now.  But strictly speaking pysinstaller does not officially support running the binary on anything other then the OS version it was compiled on.

As far as security, eh, some of us have to run older versions of OSX, no choice.

In any case I have working 10.9 binary that I built myself.  Thanks for looking into it.  I agree, its a complicated issue, if I were in your shoes I probably wouldn’t provide a binary at all given what I know now about pyinstaller.  FWIW I wouldn’t do it for linux either, which strictly speaking has the same compatibility issue waiting to pop up at some point, though linux compatibility is undoubtedly less complicated then OSX.

I would still recommend you add a manual build script to your source so that anyone installing borg from git can build the binary for their platform without vagrant.  It worked like a charm for me, but I had to manually tweak the borg.exe.spec script for it to work without a vagrant env path in there.  In theory, that change would benefit everyone on all platforms to build a binary for whatever machine they are using.  The downside is they have to also install python first in order to build it, along with the other noted dependencies.
